transport/src/main/java/io/netty/channel/IoExecutionContext.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+/**
+ * The execution context for an {@link IoHandler}.
+ * All method must be called from the {@link EventLoop} thread.
+ */
+public interface IoExecutionContext {
+    /**
+     * Returns {@code true} if blocking for IO is allowed or if we should try to do a non-blocking request for IO to be
+     * ready.
+     */
+    boolean isBlockingAllowed();
```

Comment:
@stef1927 I am not sure I like the name... suggestions welcome

Reply:
@normanmaurer while it cannot be directly applied I'd like to share our approach which may help with naming / shaping the API.

```java
public class BlockingIODetector {

    private static final ThreadLocal<Boolean> preventBlockingIO = new ThreadLocal<Boolean>() {
        @Override
        protected Boolean initialValue() {
            return Boolean.FALSE;
        }
    };

    public static void ignoreBlockingIOinThisThread() {
        preventBlockingIO.set(Boolean.FALSE);
    }
    
    public static void preventBlockingIOinThisThread() {
        preventBlockingIO.set(Boolean.TRUE);
    }

    public static void beforeBlockingIO(Method method) {
        if (preventBlockingIO.get()) {
            preventBlockingIO(method);
        }
    }

    private static void preventBlockingIO(Method method) {
        throw new UnsupportedOperationException("Call to blocking I/O " + method + " in this thread is not allowed.");
    }
    ...
}
```

Reply:
For the method name in the context, perhaps: `canBlock`, `mayBlock` or even simply `blocking` or  `isBlocking`. 

I was also thinking of having two methods in the handler, for example `runOnce` and `runBlocking` as an alternative.  However, there is a delicate interaction with the wakeup updater in the epoll handler, which I had previously failed to consider. `isBlockingAllowed` needs to be checked again after the wakeup updater has been reset to prevent user tasks from waiting until the epoll wait returns. 

The wakeup updater in the epoll wakeup method may cause us problems because we have a monitoring thread that periodically checks the queue's status of our loops, and wakes up those with tasks.

It may be that we will require a plain wakeup call that is not subject to the status of the wakeup updater. I'm not sure yet, but if we do, would it be OK for us to open a pull request later on? cc @sbtourist .

Reply:
@stef1927 yeah a follow-up PR is fine if needed

Reply:
@stef1927 will call it `canBlock()`.